stdlib/public/Concurrency/AsyncFailableFilterSequence.swift

```diff
@@ -0,0 +1,77 @@
+////===----------------------------------------------------------------------===//
+////
+//// This source file is part of the Swift.org open source project
+////
+//// Copyright (c) 2021 Apple Inc. and the Swift project authors
+//// Licensed under Apache License v2.0 with Runtime Library Exception
+////
+//// See https://swift.org/LICENSE.txt for license information
+//// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+////
+////===----------------------------------------------------------------------===//
+
+import Swift
+
+extension AsyncSequence {
+  @inlinable
+  public __consuming func filter(
+    _ isIncluded: @escaping (Element) async throws -> Bool
+  ) -> AsyncFailableFilterSequence<Self> {
+    return AsyncFailableFilterSequence(self, isIncluded: isIncluded)
+  }
+}
+
+@frozen
+public struct AsyncFailableFilterSequence<Upstream: AsyncSequence> {
+  public let upstream: Upstream
+  public let isIncluded: (Element) async throws -> Bool
+
+  @inlinable
+  public init(
+    _ upstream: Upstream, 
+    isIncluded: @escaping (Upstream.Element) async throws -> Bool
+  ) {
+    self.upstream = upstream
+    self.isIncluded = isIncluded
+  }
+}
+
+extension AsyncFailableFilterSequence: AsyncSequence {
+  public typealias Element = Upstream.Element
+  public typealias AsyncIterator = Iterator
+
+  @frozen
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var upstreamIterator: Upstream.AsyncIterator
+
+    @usableFromInline
+    let isIncluded: (Upstream.Element) async throws -> Bool
+
+    @usableFromInline
+    init(
+      _ upstreamIterator: Upstream.AsyncIterator,
+      isIncluded: @escaping (Upstream.Element) async throws -> Bool
+    ) {
+      self.upstreamIterator = upstreamIterator
+      self.isIncluded = isIncluded
+    }
+
+    @inlinable
+    public mutating func next() async throws -> Upstream.Element? {
+      while true {
+        guard let element = try await upstreamIterator.next() else {
```

Comment:
I think there is an inconsistency here with other operators on clearing the closure.

Reply:
changed to a flag